crates/optimism/rpc/src/eth/mod.rs

```diff
@@ -7,30 +7,53 @@ mod block;
 mod call;
 mod pending_block;
 
-use std::{future::Future, sync::Arc};
+use std::{fmt, sync::Arc};
 
-use alloy_primitives::{Address, U64};
-use reth_chainspec::{ChainInfo, ChainSpec};
-use reth_errors::RethResult;
+use alloy_primitives::U256;
+use derive_more::Deref;
 use reth_evm::ConfigureEvm;
-use reth_node_api::{BuilderProvider, FullNodeComponents};
-use reth_provider::{BlockReaderIdExt, ChainSpecProvider, HeaderProvider, StateProviderFactory};
-use reth_rpc::eth::DevSigner;
+use reth_network::NetworkHandle;
+use reth_network_api::NetworkInfo;
+use reth_node_api::{BuilderProvider, FullNodeComponents, FullNodeTypes};
+use reth_provider::{
+    BlockIdReader, BlockNumReader, BlockReaderIdExt, ChainSpecProvider, HeaderProvider,
+    StateProviderFactory,
+};
+use reth_rpc::eth::{core::EthApiInner, DevSigner};
 use reth_rpc_eth_api::{
     helpers::{
-        AddDevSigners, EthApiSpec, EthFees, EthSigner, EthState, LoadFee, LoadState, SpawnBlocking,
-        Trace, UpdateRawTxForwarder,
+        AddDevSigners, EthApiSpec, EthFees, EthState, LoadBlock, LoadFee, LoadState, SpawnBlocking,
+        Trace,
     },
-    EthApiTypes, RawTransactionForwarder,
+    EthApiTypes,
+};
+use reth_rpc_eth_types::{EthStateCache, FeeHistoryCache, GasPriceOracle};
+use reth_tasks::{
+    pool::{BlockingTaskGuard, BlockingTaskPool},
+    TaskExecutor, TaskSpawner,
 };
-use reth_rpc_eth_types::EthStateCache;
-use reth_rpc_types::SyncStatus;
-use reth_tasks::{pool::BlockingTaskPool, TaskSpawner};
 use reth_transaction_pool::TransactionPool;
-use tokio::sync::{AcquireError, OwnedSemaphorePermit};
 
 use crate::OpEthApiError;
 
+/// Adapter for [`EthApiInner`], which holds all the data required to serve core `eth_` API.
+pub type EthApiHandles<N> = EthApiInner<
+    <N as FullNodeTypes>::Provider,
+    <N as FullNodeComponents>::Pool,
+    NetworkHandle,
+    <N as FullNodeComponents>::Evm,
+>;
+
+/// Adapter for [`EthApiBuilderCtx`].
+pub type EthApiBuilderCtx<N> = reth_rpc_eth_types::EthApiBuilderCtx<
+    <N as FullNodeTypes>::Provider,
+    <N as FullNodeComponents>::Pool,
+    <N as FullNodeComponents>::Evm,
+    NetworkHandle,
+    TaskExecutor,
+    <N as FullNodeTypes>::Provider,
```

Comment:
now your pr about the AT network type makes perfect sense to me,

perhaps we start by removing the taskexecutor generic and replace with `Box<dyn>`? don't think the generic is actually useful

Reply:
actually, I'd like to replace both consensus trait object and task executor trait object with generics eventually. trait objects are great for things that aren't in scope for the whole program, but for consensus and task executor that isn't the case and we can make the smol perf gain by resolving them at compile time. also because we already have a great way to pass the components down to the `FullNode` type, which is via the `FullNodeComponents` trait.

made an issue https://github.com/paradigmxyz/reth/issues/9872, will tackle it before continuing with https://github.com/paradigmxyz/reth/pull/9774